[DCP][Prototype] Checkpoint replication via PGTransport

meetv18 opened the pull request:
Summary:

PR Context: Introduce simple replication logic via PGTransport. The goal is to showcase a working prototype of replication via PGTransport, in this impl we assume world_sizes are equal allowing us to create perfect bi-directional pairs for the purpose of choosing replica "partners".

Test Plan:
CI

Rollback Plan:

Differential Revision: D78034076




cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

H-Huang commented on torch/distributed/checkpoint/staging.py:
```diff
@@ -313,3 +317,76 @@ def synchronize_staging(self) -> None:
+        state_dict with its partner at (rank + world_size//2) % world_size.
+        Uses simple rank-based ordering to prevent deadlocks.
+
+        Assumes world_size is always even.
```
Add validation to raise error if not even?

H-Huang commented on test/distributed/checkpoint/test_state_dict_stager.py:
```diff
@@ -818,5 +827,424 @@ def test_dtensor(self):
+        stager.close()
+
+    @with_comms
+    def test_replication_sharded_tensors(self):
```
Thanks for these test pointers. I'll try to get sharded tensor working. If you want to comment out these tests for now in order to land, feel free

H-Huang reviewed (approved):
LGTM!

saumishr commented on torch/distributed/checkpoint/staging.py:
```diff
@@ -315,3 +321,144 @@ def synchronize_staging(self) -> None:
+
+class _ReplicationStager(AsyncStager):
+    """
+    An AsyncStager implementation that replicates state_dict across training ranks
```
Add a disclaimer that this is an experimental class and is subject to change.

meetv18 committed:
[DCP][Prototype] Checkpoint replication via PGTransport (#157963)

Summary:

### PR Context

Introduce simple replication logic via PGTransport. The goal is to showcase a working prototype of replication via PGTransport, in this impl we assume world_sizes are equal allowing us to create perfect bi-directional pairs for the purpose of choosing replica "partners".

Test Plan:
CI

Rollback Plan:

Reviewed By: saumishr, H-Huang

Differential Revision: D78034076

yangw-dev commented:
@pytorchbot merge
